EIPS/pairings.md

```diff
@@ -0,0 +1,141 @@
+  10857046999023057135944570762232829481370756359578518086990519993285655852781,
+  4082367875863433681332203403145435568316851327593401208105741076214120093531 * i +
+  8495653923123431417604973247489272438418190587263600148770280649306958101930
+)
+```
+
+
+### Encoding
+
+Elements of `F_p` are encoded as 32 byte big-endian numbers. An encoding value of `p` or larger is invalid.
+
+Elements `a * i + b` of `F_p^2` are encoded as two elements of `F_p`, `(a, b)`.
+
+Elliptic curve points are encoded as a Jacobian pair `(X, Y)` where the point at infinity is encoded as `(0, 0)`.
+
+Note that the number `k` is derived from the input length.
+
+The length of the returned data is always exactly 32 bytes and encoded as a 32 byte big-endian number. 
+
+### Gas costs
+
+[Benchmarks run on cpp-ethereum](https://gist.github.com/chriseth/4168b56bfe638cae8da1945dd988600b)
+
+suggest the following gas formula:
+
+`60000 * k + 40000`
+
+if we target 20000 gas per millisecond.
+
+Awaiting benchmarks from other implementations.
+
+## Rationale
+
+The specific curve `alt_bn128` was chosen because it is particularly well-suited for zkSNARKs, or, more specifically their verification building block of pairing functions. Furthermore, by choosing this curve, we can use synergy effects with ZCash and re-use some of their components and artifacts.
+
+The feature of adding curve and field parameters to the inputs was considered but ultimately rejected since it complicates the specification: The gas costs are much harder to determine and it would be possible to call the contracts on something which is not an actual elliptic curve or does not admit an efficient pairing implementation.
+
+A non-compact point encoding was chosen since it still allows to perform some operations in the smart contract itself (inclusion of the full y coordinate) and two encoded points can be compared for equality (no third projective coordinate).
+
+The encoding of field elements in `F_p^2` was chosen in this order to be in line with the big endian encoding of the elements themselves.
+
+## Backwards Compatibility
+
+As with the introduction of any precompiled contract, contracts that already use the given addresses will change their semantics. Because of that, the addresses are taken from the "reserved range" below 256.
+
+## Test Cases
+
+To be written.
+
+## Implementation
+
+The precompiled contract can be implemented using elliptic curve pairing functions, more specifically, an optimal ate pairing on the alt_bn128 curve, which can be implemented efficiently. In order to see that, first note that a pairing function `e: G_1 x G_2 -> G_T` fulfills the following properties (`G_1` and `G_2` are written additively, `G_T` is written multiplicatively):
+
+(1) `e(m * P1, n * P2) = e(P1, P2)^(m * n)`
+(2) `e` is non-degenerate
+
+Now observe that
+```
+log_P1(a1) * log_P2(b1) + ... + log_P1(ak) * log_P2(bk) = 0
+```
+if and only if
```

Comment:
This "if and only if" is broken as [pointed out](https://github.com/ethereum/EIPs/pull/212#issuecomment-316150345) by @bbuenz .